pytential/symbolic/pde/system_utils.py

```diff
@@ -0,0 +1,529 @@
+            expr = expr.diff(ds[knl.axis])
+            found = True
+        else:
+            return [int_g]
+        knl = knl.inner_kernel
+
+    if not found:
+        return [int_g]
+
+    int_g = int_g.copy(target_kernel=knl)
+
+    sources_pymbolic = [NodeCoordinateComponent(i) for i in range(knl.dim)]
+    expr = expr.expand()
+    # Now the expr is an Add and looks like
+    # u_{d[0], d[1]}(d, y)*d[0]*y[1] + u(d, y) * d[1]
+    # or a single term like u(d, y) * d[1]
+    if isinstance(expr, sympy.Add):
+        args = expr.args
+    else:
+        args = [expr]
+
+    result = []
+    for arg in args:
+        deriv_terms = arg.atoms(sympy.Derivative)
+        if len(deriv_terms) == 1:
+            # for eg: we have a term like u_{d[0], d[1]}(d, y) * d[0] * y[1]
+            # deriv_term is u_{d[0], d[1]}
+            deriv_term = deriv_terms.pop()
+            # eg: rest_terms is d[0] * y[1]
+            rest_terms = sympy.Poly(arg.xreplace({deriv_term: 1}), *ds, *sources)
+            # eg: derivatives is (d[0], 1), (d[1], 1)
+            derivatives = deriv_term.args[1:]
+        elif len(deriv_terms) == 0:
+            # for eg: we have a term like u(d, y) * d[1]
+            # rest_terms = d[1]
+            rest_terms = sympy.Poly(arg.xreplace({orig_expr: 1}), *ds, *sources)
+            derivatives = [(d, 0) for d in ds]
+        else:
+            raise AssertionError("impossible condition")
+
+        # apply the derivatives
+        new_source_kernels = []
+        for source_kernel in int_g.source_kernels:
+            knl = source_kernel
+            for axis_var, nrepeats in derivatives:
+                axis = ds.index(axis_var)
+                for _ in range(nrepeats):
+                    knl = AxisSourceDerivative(axis, knl)
+            new_source_kernels.append(knl)
+        new_int_g = int_g.copy(source_kernels=new_source_kernels)
+
+        (monom, coeff,) = rest_terms.terms()[0]
+        # Now from d[0]*y[1], we separate the two terms
+        # d terms end up in the expression and y terms end up in the density
+        d_terms, y_terms = monom[:len(ds)], monom[len(ds):]
+        expr_multiplier = _monom_to_expr(d_terms, ds)
+        density_multiplier = _monom_to_expr(y_terms, sources_pymbolic) \
+                * conv(coeff)
+        # since d/d(d) = - d/d(y), we multiply by -1 to get source derivatives
+        density_multiplier *= (-1)**int(sum(nrepeats for _, nrepeats in derivatives))
+        new_int_gs = _multiply_int_g(new_int_g, sym.sympify(expr_multiplier),
+                density_multiplier)
+        result.extend(new_int_gs)
+    return result
+
+
+def _multiply_int_g(int_g: IntG, expr_multiplier: sym.Basic,
+        density_multiplier: ExpressionT) -> List[IntG]:
+    """Multiply the expression in ``IntG`` with the *expr_multiplier*
+    which is a symbolic expression and multiply the densities
+    with *density_multiplier* which is a pymbolic expression.
```

Comment:
```suggestion
    which is a :mod:`sympy` expression and multiply the densities
    with *density_multiplier* which is a :mod:`pymbolic` expression.
```
?

Reply:
Fixed in f4032b72188a5c48670878a7e3249ec6e02a8167